homeassistant/components/scheduler/schedule.py

```diff
@@ -0,0 +1,371 @@
+DOMAIN = "scheduler"
+
+STORAGE_VERSION = 1
+
+CONF_END_DATETIME = "end_datetime"
+CONF_RECURRENCE = "recurrence"
+CONF_SCHEDULE_ID = "schedule_id"
+CONF_START_DATETIME = "start_datetime"
+
+STATE_EXPIRED = "expired"
+
+TOPIC_SCHEDULE_NEXT = "scheduler_schedule_next"
+
+
+class ScheduleInstance:
+    """A class to represent an active instance of a schedule."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entity_id: str,
+        start_datetime: datetime,
+        end_datetime: Optional[datetime] = None,
+    ):
+        """Initialize."""
+        self._affected_states: List[State] = []
+        self._async_revert_listener: Optional[Callable[..., Awaitable]] = None
+        self._async_state_listener: Optional[Callable[..., Awaitable]] = None
+        self._async_trigger_listener: Optional[Callable[..., Awaitable]] = None
+        self._context: Context = Context()
+        self._hass: HomeAssistant = hass
+        self.end_datetime: Optional[datetime] = end_datetime
+        self.entity_id: str = entity_id
+        self.start_datetime: datetime = start_datetime
+
+    @callback
+    def async_init(self) -> None:
+        """Set up scheduling for this instance."""
+
+        async def revert(executed_at: datetime) -> None:
+            """Revert."""
+            await self.async_revert()
+
+        async def trigger(executed_at: datetime) -> None:
+            """Trigger."""
+            await self.async_trigger()
+
+        self._async_trigger_listener = async_track_point_in_time(
+            self._hass, trigger, self.start_datetime
+        )
+
+        if self.end_datetime:
+            self._async_trigger_listener = async_track_point_in_time(
+                self._hass, revert, self.end_datetime
+            )
+
+    @callback
+    def async_cancel(self) -> None:
+        """Cancel the instance."""
+        if self._async_revert_listener:
+            self._async_revert_listener()
+            self._async_revert_listener = None
+        if self._async_state_listener:
+            self._async_state_listener()
+            self._async_state_listener = None
+        if self._async_trigger_listener:
+            self._async_trigger_listener()
+            self._async_trigger_listener = None
+
+    async def async_revert(self) -> None:
+        """Restore the entities touched by the schedule."""
+        if not self._affected_states:
+            return
+
+        entities = {}
+        for state in self._affected_states:
+            data = {**state.attributes}
+            data["state"] = state.state
+            entities[state.entity_id] = data
+
+        await self._hass.services.async_call(
+            "scene",
+            "apply",
+            service_data={"entities": entities},
+            blocking=True,
+            context=self._context,
+        )
+
+        _LOGGER.info("Scheduler reverted scene: %s", self.entity_id)
+
+        async_dispatcher_send(self._hass, TOPIC_SCHEDULE_NEXT)
```

Comment:
@balloob Thoughts on this? By moving the listener management into the instance, I lost the ability for `Schedule` to automatically schedule the next instance once the original one is complete. This works, but is it right?